services/brig/src/Brig/Data/Activation.hs

```diff
@@ -135,6 +137,17 @@ activateKey k c u = verifyCode k c >>= pickUser >>= activate
         updateEmailAndDeleteEmailUnvalidated :: MonadClient m => UserId -> Email -> m ()
         updateEmailAndDeleteEmailUnvalidated u' email =
           updateEmail u' email <* deleteEmailUnvalidated u'
+        deleteCode :: UserId -> m ()
+        deleteCode uId =
+          runM
+            -- FUTUREWORK: use the DeletePasswordResetCode action instead of CodeDelete
+            ( codeStoreToCassandra @m
+                ( embed @m
+                    ( liftIO @m (runM (passwordResetSupplyToIO @'[Embed IO] (E.mkPasswordResetKey uId)))
```

Comment:
This looks wrong to me. It should be something like

```
runM . codeStoreToCassandra . passwordResetSupplyToIO $ E.mkPasswordResetKey uId >>= E.codeDelete
```

Reply:
But also this is a dangerous thing to do in general and would be better to push these constraints up to the type signature instead.

Reply:
> But also this is a dangerous thing to do in general and would be better to push these constraints up to the type signature instead.

The function doesn't yet run in `Sem`, hence this soon-to-be-done-away-with running of effects. I will push the constraints up to the type signature in one of the follow-up PRs.

Reply:
@isovector , in case you find time, let me know what you think about what I wrote above.

Reply:
I had another look at this, and I think I understand the issue you're running into. This function doesn't yet run in `Sem`, but you want to run `Sem` actions inside of it?

The solution here is to port this function into `Sem`. To do the `runM` dance currently in this PR is extremely dangerous. By using `runM`, you are committing to a particular interpretation of your effects, which may or *may not* be the same choices that are made globally. That means, eg, maybe a test wants to run against an in memory database, but this function has committed to using Cassandra. In the best case, your tests will fail, in the worst, you'll corrupt the database invariants.

I'd recommend using this technique even in the short-term as a crutch, because the compiler can't help if we forget to take the crutch out. It'll just be a sitting time bomb, waiting to waste someone's day when they run afoul of the presumed interpretation.

Reply:
A better solution here for the interim is to give (orphan) `Member (Embed Client) r => MonadClient (Sem r)` and similar `MonadReader` instances to `Sem`. Then you don't need to play the interpretation game, and you can still interop mtl code with polysemy before having the whole thing ported.